File: v1/packages/core/poseAnalyzer.ts
// Ported from v0/backend/app/services/pose_analyzer.py
// TypeScript version for RoboQoach v1
const squatSpecs = require('./exercise_specs.json');

export type Keypoint = [number, number, number]; // [x, y, confidence]
export type Keypoints = Keypoint[];

const KEYPOINT_MAP: { [key: string]: number } = {
  nose: 0, left_eye: 1, right_eye: 2, left_ear: 3, right_ear: 4,
  left_shoulder: 5, right_shoulder: 6, left_elbow: 7, right_elbow: 8,
  left_wrist: 9, right_wrist: 10, left_hip: 11, right_hip: 12,
  left_knee: 13, right_knee: 14, left_ankle: 15, right_ankle: 16
};

const GENERIC_TO_KEYPOINT: { [key: string]: string } = {
  shoulder: 'left_shoulder',
  elbow: 'left_elbow',
  wrist: 'left_wrist',
  hip: 'left_hip',
  knee: 'left_knee',
  ankle: 'left_ankle',
};

function pointFromName(name: string, keypoints: Keypoints): Keypoint {
  const mapped = GENERIC_TO_KEYPOINT[name] || name;
  const idx = KEYPOINT_MAP[mapped];
  if (idx === undefined || idx >= keypoints.length) return [0, 0, 0];
  return keypoints[idx];
}

function angle(a: Keypoint, b: Keypoint, c: Keypoint): number {
  const ba = [a[0] - b[0], a[1] - b[1]];
  const bc = [c[0] - b[0], c[1] - b[1]];
  const denom = Math.hypot(...ba) * Math.hypot(...bc);
  if (denom === 0) return 0;
  const cosVal = Math.max(-1, Math.min(1, (ba[0] * bc[0] + ba[1] * bc[1]) / denom));
  return Math.acos(cosVal) * (180 / Math.PI);
}

function horizontalDeviation(p1: Keypoint, p2: Keypoint): number {
  return Math.abs(p1[0] - p2[0]);
}

function verticalAlignment(points: Keypoint[]): number {
  const xs = points.filter(p => p[2] > 0.3).map(p => p[0]);
  if (xs.length < 2) return 1.0;
  const mean = xs.reduce((a, b) => a + b, 0) / xs.length;
  const variance = xs.reduce((a, b) => a + (b - mean) ** 2, 0) / xs.length;
  return Math.sqrt(variance);
}

export function analyzePose(exerciseType: string, keypoints: Keypoints) {
  const et = exerciseType.toLowerCase();
  const spec = squatSpecs[et];
  if (!spec) {
    return {
      feedback: `${exerciseType} analysis coming soon!`,
      corrections: ['Keep good form'],
      rep_count: 0,
      is_good_rep: false
    };
  }
  const results = [];
  const corrections: string[] = [];
  let good = true;
  for (const rule of spec.rules || []) {
    const rtype = rule.type;
    const pts = rule.points.map((n: string) => pointFromName(n, keypoints));
    let ok = true;
    let value: number | undefined = undefined;
    if (rtype === 'angle' && pts.length === 3) {
      value = angle(pts[0], pts[1], pts[2]);
      const min_v = rule.min ?? -9999;
      const max_v = rule.max ?? 9999;
      ok = value >= min_v && value <= max_v;
    } else if (rtype === 'horizontal_distance' && pts.length === 2) {
      value = horizontalDeviation(pts[0], pts[1]);
      const maxDev = rule.maxDeviation ?? 0.2;
      ok = value <= maxDev;
    } else if (rtype === 'vertical_alignment' && pts.length >= 2) {
      value = verticalAlignment(pts);
      const maxDev = rule.maxDeviation ?? 0.2;
      ok = value <= maxDev;
    }
    results.push({
      id: rule.id,
      type: rtype,
      value,
      ok,
      severity: rule.severity || 'low'
    });
    if (!ok) {
      corrections.push(rule.errorMessage || 'Adjust your form');
      if (rule.severity === 'high') good = false;
    }
  }
  // Rep detection
  const rep_cfg = spec.repDetection || {};
  const lm = rep_cfg.landmark;
  const axis = rep_cfg.axis || 'y';
  const threshold = rep_cfg.threshold || 0.5;
  const dirn = rep_cfg.direction || 'down_up';
  let rep_value: number | undefined = undefined;
  if (lm) {
    const p = pointFromName(lm, keypoints);
    rep_value = axis === 'y' ? p[1] : p[0];
  }
  const is_good_rep = good && results.every(r => r.ok);
  const feedback = is_good_rep ? 'Excellent form!' : corrections[0] || 'Keep steady';
  return {
    feedback,
    corrections,
    rules: results,
    rep_signal: {
      landmark: lm,
      axis,
      value: rep_value,
      threshold,
      direction: dirn
    },
    rep_count: 0,
    is_good_rep
  };
}

export function calculateFormScore(exerciseType: string, keypoints: Keypoints): number {
  const visible = keypoints.filter(kp => kp[2] > 0.5);
  if (visible.length === 0) return 0.0;
  const avgConfidence = visible.reduce((a, b) => a + b[2], 0) / visible.length;
  return Math.round(avgConfidence * 1000) / 10;
}

File: v0/backend/app/services/pose_analyzer.py
import json
import os
import numpy as np
from typing import Dict, List, Any, Tuple

# MoveNet keypoint indices
KEYPOINT_MAP = {
    'nose': 0, 'left_eye': 1, 'right_eye': 2, 'left_ear': 3, 'right_ear': 4,
    'left_shoulder': 5, 'right_shoulder': 6, 'left_elbow': 7, 'right_elbow': 8,
    'left_wrist': 9, 'right_wrist': 10, 'left_hip': 11, 'right_hip': 12,
    'left_knee': 13, 'right_knee': 14, 'left_ankle': 15, 'right_ankle': 16
}

# Map generic names to left-side landmarks by default
GENERIC_TO_KEYPOINT = {
    'shoulder': 'left_shoulder',
    'elbow': 'left_elbow',
    'wrist': 'left_wrist',
    'hip': 'left_hip',
    'knee': 'left_knee',
    'ankle': 'left_ankle',
}

def _load_specs() -> Dict[str, Any]:
    here = os.path.dirname(__file__)
    path = os.path.join(here, 'exercise_specs.json')
    try:
        with open(path, 'r') as f:
            return json.load(f)
    except Exception as e:
        print('Failed to load exercise specs:', e)
        return {}

SPECS = _load_specs()

def _point_from_name(name: str, keypoints: np.ndarray) -> Tuple[float, float, float]:
    mapped = GENERIC_TO_KEYPOINT.get(name, name)
    idx = KEYPOINT_MAP.get(mapped)
    if idx is None or idx >= keypoints.shape[0]:
        return (0.0, 0.0, 0.0)
    return tuple(keypoints[idx])  # (x, y, c)

def _angle(a, b, c) -> float:
    # angle at b between ba and bc in degrees
    ba = np.array([a[0]-b[0], a[1]-b[1]])
    bc = np.array([c[0]-b[0], c[1]-b[1]])
    denom = (np.linalg.norm(ba) * np.linalg.norm(bc))
    if denom == 0:
        return 0.0
    cos_val = np.clip(np.dot(ba, bc) / denom, -1.0, 1.0)
    return float(np.degrees(np.arccos(cos_val)))

def _horizontal_deviation(p1, p2) -> float:
    # normalized by image width (assumes x in [0,1])
    return abs(p1[0] - p2[0])

def _vertical_alignment(points: List[Tuple[float, float, float]]) -> float:
    # std dev of x positions as a proxy for vertical alignment; lower is better
    xs = [p[0] for p in points if p[2] > 0.3]
    if len(xs) < 2:
        return 1.0
    return float(np.std(xs))

def analyze_pose(exercise_type: str, keypoints: np.ndarray) -> Dict[str, Any]:
    """Analyze pose based on exercise type and return feedback using JSON-driven rules"""
    et = exercise_type.lower()
    spec = SPECS.get(et)
    if not spec:
        return {
            "feedback": f"{exercise_type} analysis coming soon!",
            "corrections": ["Keep good form"],
            "rep_count": 0,
            "is_good_rep": False
        }

    results = []
    corrections: List[str] = []
    good = True

    for rule in spec.get('rules', []):
        rtype = rule.get('type')
        pts = [ _point_from_name(n, keypoints) for n in rule.get('points', []) ]
        ok = True
        value = None

        if rtype == 'angle' and len(pts) == 3:
            value = _angle(pts[0], pts[1], pts[2])
            min_v = rule.get('min', -9999)
            max_v = rule.get('max', 9999)
            ok = (value >= min_v) and (value <= max_v)
        elif rtype == 'horizontal_distance' and len(pts) == 2:
            value = _horizontal_deviation(pts[0], pts[1])
            max_dev = rule.get('maxDeviation', 0.2)
            ok = value <= max_dev
        elif rtype == 'vertical_alignment' and len(pts) >= 2:
            value = _vertical_alignment(pts)
            max_dev = rule.get('maxDeviation', 0.2)
            ok = value <= max_dev
        else:
            ok = True  # unknown rule types considered pass

        results.append({
            'id': rule.get('id'),
            'type': rtype,
            'value': value,
            'ok': ok,
            'severity': rule.get('severity', 'low')
        })

        if not ok:
            corrections.append(rule.get('errorMessage', 'Adjust your form'))
            if rule.get('severity') == 'high':
                good = False

    # Simple rep position signal
    rep_cfg = spec.get('repDetection', {})
    lm = rep_cfg.get('landmark')
    axis = rep_cfg.get('axis', 'y')
    threshold = rep_cfg.get('threshold', 0.5)
    dirn = rep_cfg.get('direction', 'down_up')
    rep_value = None
    if lm:
        p = _point_from_name(lm, keypoints)
        coord = p[1] if axis == 'y' else p[0]
        rep_value = float(coord)
    is_good_rep = good and all(r['ok'] for r in results)

    feedback = "Excellent form!" if is_good_rep else (corrections[0] if corrections else "Keep steady")

    return {
        'feedback': feedback,
        'corrections': corrections,
        'rules': results,
        'rep_signal': {
            'landmark': lm,
            'axis': axis,
            'value': rep_value,
            'threshold': threshold,
            'direction': dirn,
        },
        'rep_count': 0,
        'is_good_rep': is_good_rep,
    }

def analyze_squat(keypoints: np.ndarray) -> Dict[str, Any]:  # backwards compatibility if used elsewhere
    return analyze_pose('squat', keypoints)

def analyze_pushup(keypoints: np.ndarray) -> Dict[str, Any]:
    return analyze_pose('pushup', keypoints)

def calculate_form_score(exercise_type: str, keypoints: np.ndarray) -> float:
    """Calculate overall form score (0-100)"""
    try:
        # Average confidence of visible keypoints
        visible_keypoints = keypoints[keypoints[:, 2] > 0.5]
        if len(visible_keypoints) == 0:
            return 0.0
        
        avg_confidence = np.mean(visible_keypoints[:, 2])
        return round(avg_confidence * 100, 1)
    except:
        return 50.0  # Default score if calculation fails
